File: bound/kt/src/main/kotlin/tbdex/sdk/http/Offerings.kt
package tbdex.sdk.http

import tbdex.sdk.TbdexException
import tbdex.sdk.resources.Offering
import tbdex.sdk.rust.GetOfferingsResponseBody as RustCoreGetOfferingsResponseBody

data class GetOfferingsResponseBody private constructor(
    val data: List<Offering>,
    internal val rustCoreGetOfferingsResponseBody: RustCoreGetOfferingsResponseBody
) {
    constructor(offerings: List<Offering>) : this(
        offerings,
        try {
            RustCoreGetOfferingsResponseBody(
                offerings.map { it.rustCoreOffering }
            )
        } catch (e: tbdex.sdk.rust.TbdexException.Exception) {
            throw TbdexException.fromRustCore(e)
        }
    )

    companion object {
        fun fromJsonString(json: String): GetOfferingsResponseBody {
            try {
                val rustCoreGetOfferingsResponseBody = RustCoreGetOfferingsResponseBody.fromJsonString(json)
                val offerings = rustCoreGetOfferingsResponseBody.getData().data.map {
                    Offering.fromRustCoreOffering(it)
                }
                return GetOfferingsResponseBody(offerings, rustCoreGetOfferingsResponseBody)
            } catch (e: tbdex.sdk.rust.TbdexException.Exception) {
                throw TbdexException.fromRustCore(e)
            }
        }
    }

    fun toJsonString(): String {
        try {
            return rustCoreGetOfferingsResponseBody.toJsonString()
        } catch (e: tbdex.sdk.rust.TbdexException.Exception) {
            throw TbdexException.fromRustCore(e)
        }
    }
}

File: bound/typescript/src/http/offerings.ts
import { Offering } from "../resources/offering";

export class GetOfferingsResponseBody {
  readonly data: Offering[];

  constructor(data: Offering[]) {
    this.data = data;
  }

  static fromJSONString = (json: string): GetOfferingsResponseBody => {
    const obj = JSON.parse(json);
    return new GetOfferingsResponseBody(
      obj.data.map(
        (x: Offering) => new Offering(x.metadata, x.data, x.signature)
      )
    );
  };

  toJSONString = (): string => {
    return JSON.stringify({ data: this.data });
  };
}
